### PR TITLE
[#320] [FEATURE] Rendre (dés-)activable les pré-traitements des réponses/solutions.

### PR DESCRIPTION
**Pull request by bdavidxyz**

### CODE REVIEW

**Comment by jbuget** [comment]
I've deployed this PR to http://330-answers-solutions-pretreatments-deactivatable.pix.beta.gouv.fr. Please check it out

**Comment by jbuget** [comment]
I've deployed this PR to http://330-answers-solutions-pretreatments-deactivatable.pix.beta.gouv.fr. Please check it out

**Comment by jbuget** [comment]
I've deployed this PR to http://330-answers-solutions-pretreatments-deactivatable.pix.beta.gouv.fr. Please check it out

**Comment by jbuget** [comment]
I've deployed this PR to http://330-answers-solutions-pretreatments-deactivatable.pix.beta.gouv.fr. Please check it out

**Comment by jbuget** [comment]
I've deployed this PR to http://330-answers-solutions-pretreatments-deactivatable.pix.beta.gouv.fr. Please check it out

**Comment by jbuget** [comment]
I've deployed this PR to http://330-answers-solutions-pretreatments-deactivatable.pix.beta.gouv.fr. Please check it out

**Comment by jbuget** [comment]
I've deployed this PR to http://330-answers-solutions-pretreatments-deactivatable.pix.beta.gouv.fr. Please check it out

**Comment by jbuget** [comment]
I've deployed this PR to http://330-answers-solutions-pretreatments-deactivatable.pix.beta.gouv.fr. Please check it out

**Comment by jbuget** [comment]
I've deployed this PR to http://330-answers-solutions-pretreatments-deactivatable.pix.beta.gouv.fr. Please check it out

**Comment by jbuget** [comment]
I've deployed this PR to http://330-answers-solutions-pretreatments-deactivatable.pix.beta.gouv.fr. Please check it out

**Comment by jbuget** [comment]
I've deployed this PR to http://330-answers-solutions-pretreatments-deactivatable.pix.beta.gouv.fr. Please check it out

**Comment by jbuget** [comment]
I've deployed this PR to http://330-answers-solutions-pretreatments-deactivatable.pix.beta.gouv.fr. Please check it out

**Comment by jbuget** [comment]
I've deployed this PR to http://330-answers-solutions-pretreatments-deactivatable.pix.beta.gouv.fr. Please check it out

**Comment by jbuget** [comment]
I've deployed this PR to http://330-answers-solutions-pretreatments-deactivatable.pix.beta.gouv.fr. Please check it out

**Comment by jbuget** [comment]
I've deployed this PR to http://330-answers-solutions-pretreatments-deactivatable.pix.beta.gouv.fr. Please check it out

**Comment by jbuget** [comment]
I've deployed this PR to http://330-answers-solutions-pretreatments-deactivatable.pix.beta.gouv.fr. Please check it out

**Comment by jbuget** [comment]
I've deployed this PR to http://330-answers-solutions-pretreatments-deactivatable.pix.beta.gouv.fr. Please check it out

**Comment by jbuget** [comment]
I've deployed this PR to http://330-answers-solutions-pretreatments-deactivatable.pix.beta.gouv.fr. Please check it out

**Comment by jbuget** [comment]
I've deployed this PR to http://330-answers-solutions-pretreatments-deactivatable.pix.beta.gouv.fr. Please check it out

**Comment by jbuget** [comment]
I've deployed this PR to http://330-answers-solutions-pretreatments-deactivatable.pix.beta.gouv.fr. Please check it out

**Comment by jbuget** [comment]
I've deployed this PR to http://330-answers-solutions-pretreatments-deactivatable.pix.beta.gouv.fr. Please check it out

**Comment by jbuget** [comment]
I've deployed this PR to http://330-answers-solutions-pretreatments-deactivatable.pix.beta.gouv.fr. Please check it out

**Comment by jbuget** [comment]
I've deployed this PR to http://330-answers-solutions-pretreatments-deactivatable.pix.beta.gouv.fr. Please check it out

**Comment by jbuget** [comment]
I've deployed this PR to http://330-answers-solutions-pretreatments-deactivatable.pix.beta.gouv.fr. Please check it out

**Comment by jbuget** [comment]
I've deployed this PR to http://330-answers-solutions-pretreatments-deactivatable.pix.beta.gouv.fr. Please check it out

**Comment by jbuget** [comment]
I've deployed this PR to http://330-answers-solutions-pretreatments-deactivatable.pix.beta.gouv.fr. Please check it out

**Comment by jbuget** [comment]
I've deployed this PR to http://330-answers-solutions-pretreatments-deactivatable.pix.beta.gouv.fr. Please check it out

**Comment by jbuget** [comment]
I've deployed this PR to http://330-answers-solutions-pretreatments-deactivatable.pix.beta.gouv.fr. Please check it out

File: api/lib/domain/services/solution-service-qroc.js [hunk 1]
```diff
@@ -8,10 +9,43 @@ function _applyPreTreatmentsToSolutions(solution) {
             .value();
 }
 
-function _applyTreatmentsToSolutions(solution) {
+function _applyTreatmentsToSolutions(solution, deactivations) {
   const pretreatedSolutions = _applyPreTreatmentsToSolutions(solution);
   return  _.map(pretreatedSolutions, (pretreatedSolution) => {
-    return utils._treatmentT2(utils._treatmentT1(pretreatedSolution));
+
+    // default behaviour : all treatments applies
```
Comment (jbuget) [inline api/lib/domain/services/solution-service-qroc.js]: Les méthodes appelées sont suffisamment bien nommées pour qu'on se passe de commentaire.
Par ailleurs, je pense que le commentaire ne décrit pas la réalité : il existe 3 traitements (T1, T2, T3). Le commentaire dit "tous les traitements", or le code affiche seulement T1 et T2.

File: api/lib/domain/services/solution-service-qroc.js [hunk 2]
```diff
@@ -21,23 +55,88 @@ function _applyAnswerTreatment(strArg) {
 }
 
 
+function _calculateResult(validations, deactivations) {
+
+  // default behaviour
```
Comment (jbuget) [inline api/lib/domain/services/solution-service-qroc.js]: idem, le code est à mon sens largement propre pour ne pas avoir besoin des commentaires qui sont redondants avec le code

File: api/lib/domain/services/solution-service-qrocm-ind.js [hunk 3]
```diff
@@ -1,11 +1,51 @@
...
+function _applyTreatmentsToSolutions(solutions, deactivations) {
   return _.mapValues(solutions, (validSolutions) => {
     return _.map(validSolutions, (validSolution) => {
-      return utils._treatmentT2(utils._treatmentT1(validSolution.toString()));
+
+      if (deactivationsService.isDefault(deactivations)) {
+        return utils._treatmentT2(utils._treatmentT1(validSolution));
+      }
+
+      // Only T1 is deactivated
```
Comment (jbuget) [inline api/lib/domain/services/solution-service-qrocm-ind.js]: idem que plus haut, le code est propre => les commentaires inutiles

File: api/lib/domain/services/solution-service-qrocm-ind.js [hunk 4]
```diff
@@ -15,13 +55,67 @@ function _applyTreatmentsToAnswers(answers) {
...
 
-function _calculateResult(validations) {
+function _calculateResult(validations, deactivations) {
   let result = 'ok';
 
   _.each(validations, (validation) => {
-    if (validation.t1t2t3Ratio > 0.25) {
-      result = 'ko';
+
+    // default behaviour : all T1, T2 and T3 are ACTIVATED
```
Comment (jbuget) [inline api/lib/domain/services/solution-service-qrocm-ind.js]: idem que plus haut, le code est propre => les commentaires inutiles

File: api/tests/unit/domain/services/solution-service-function-timedOut_test.js [hunk 5]
```diff
@@ -0,0 +1,38 @@
+const { describe, it, expect } = require('../../../test-helper');
+
+const service = require('../../../../lib/domain/services/solution-service');
+
+describe('Unit | Service | SolutionService', function () {
+
+  describe('#_timedOut', function () {
+    it('should return "timedout" if result is partially correct and timeout is negative', function () {
```
Comment (jbuget) [inline api/tests/unit/domain/services/solution-service-function-timedOut_test.js]: je pense qu'un bon gros `.forEach` pourrait alléger et simplifier le code

**Review by jbuget — APPROVED** [review]
Des commentaires en trop mais sinon c'est bon pour moi. Je mets l'étiquette "changes required" juste pour qu'on n'oublie pas de les retirer. @bdavidxyz tu peux mettre le label "LGT Jérémy" et merger dès que c'est pris en compte. Bravo et merci pour TOUS ces tests :-)

**Comment by jbuget** [comment]
I've deployed this PR to http://330-answers-solutions-pretreatments-deactivatable.pix.beta.gouv.fr. Please check it out

File: api/lib/domain/services/solution-service-qroc.js (hunk 1 above)
Comment (bdavidxyz) [reply api/lib/domain/services/solution-service-qroc.js]: ok je supprime les commentaires

File: api/tests/unit/domain/services/solution-service-function-timedOut_test.js (hunk 5 above)
Comment (bdavidxyz) [reply api/tests/unit/domain/services/solution-service-function-timedOut_test.js]: ok je m'en occupe

**Comment by jbuget** [comment]
I've deployed this PR to http://330-answers-solutions-pretreatments-deactivatable.pix.beta.gouv.fr. Please check it out

**Comment by jbuget** [comment]
I've deployed this PR to http://330-answers-solutions-pretreatments-deactivatable.pix.beta.gouv.fr. Please check it out

**Comment by jbuget** [comment]
I've deployed this PR to http://330-answers-solutions-pretreatments-deactivatable.pix.beta.gouv.fr. Please check it out

**Pull request merged**